# SUSA CX Kaggle Capstone
## Part 1: Introduction, Exploratory Data Analysis, and Feature Selection
### Hosted by and maintained by the [Statistics Undergraduate Students Association (SUSA)](https://susa.berkeley.edu). Originally authored by [Arun Ramamurthy](mailto:contact@arun.run), [Patrick Chao](mailto:prc@berkeley.edu), & [Noah Gundotra](mailto:noah.gundotra@berkeley.edu).


# SUSA CX Kaggle Capstone Project

Welcome to the last four weeks of your semester in SUSA's Career Exploration committee! Now that you've participated in nearly a dozen workshops on Python, R, data science, and machine learning, we're going to guide you through a four-week collaborative Kaggle competition with your peers in Career Exploration. We want to give you the experience of working with real data, using real machine learning algorithms, in an educational setting. You will have to make use of your data computing skills in Python, dive into reading kernels on the Kaggle website, use visualization and feature engineering to improve your score, and maybe even pick up a few advanced deep learning models along the way. If this sounds a bit intimidating right now, do not fret! Your SUSA Mentors will be there to mentor you through the whole thing. So without further ado, let's get cracking!

## What is Kaggle?

### Kaggle Competition: Housing Prices

In this project, we will be working with the [**Housing Prices**](https://www.kaggle.com/c/house-prices-advanced-regression-techniques/) competition on the Kaggle website. The competition deals with accurately predicting the sales prices of houses in Ames, Iowa. More information on the dataset and competition will follow shortly.

## Machine Learning

## Accessing the `House Prices` Dataset

While you can access these datasets online at the [Kaggle webpage for this competition](https://www.kaggle.com/c/house-prices-advanced-regression-techniques/data), we have already downloaded the data for you and placed it in the `crash-course/Kaggle/DATA/house-prices/` directory. 

## Logistics

This project is going to be tough and long, but we will be there with you through the whole thing. Additionally, you won't be working alone - we have used your input to assign you into teams of 3 - 4 members. These will be your team members for the rest of the semester, so get to know each other a little!   

The teams are as follows:


# The Data Science Workflow

In general, there are a few key steps to begin working with a dataset. 

First, we need to **understand what the dataset actually is about**, and what we are trying to do with it. Key to this stage is understanding what each row of the dataset represents, as well as what each column indicates. You can read more about this dataset by looking at the dataset itself, or reading the data dictionary in `crash-course/Kaggle/DATA/house-prices/data_description.txt`.

Second, before we can even do exploratory data analysis, we need to **clean our dataset**. It is very helpful to identify the size of the dataset, so we know how many samples we have. We need to determine a consistent method of dealing with missing values, such as setting them to a value, removing the feature entirely, interpolating values, etc. Other crucial steps are separating into training and validation, as well as creating elementary data plots.

Third, we have **exploratory data analysis (EDA)**. The point of this phase is to inspect & visualize key relationships, trends, outliers, and issues with our data. By conducting EDA, we get a better sense of the underlying structure of our data, and which features are most important. Especially since we have 81 features, we would like to select the features that are most important to our analysis before we begin modeling `SalePrice`. Once we have an idea about which features to use and how they relate to one another, our modeling stage will be more informed and robust. 

Fourth, we have the **modeling phase**, consisting of **model selection** and **model training**. Here, we select a predictive model to train on our features, and then actually train the model! In this first week, we will be using linear regression as a first-pass. In later weeks of the SUSA CX Kaggle Capstone project, we will be using more advanced models like random forest and neural networks. Depending on the model we are using, it is important to verify the model's assumptions before fully pledging to that model. Additionally, we may use **validation** in this stage to select certain hyperparameters for our model selection.

Finally, we have the **model evaluation phase**. Here, we compute a metric for our model's performance, usually by  summing the squared errors of the model's predictions on the test set. This stage allows us to effectively compare various data cleaning and modeling selection decisions, by giving us a single comparable value for performance across our potential models.

## I. Understanding our Dataset 
Our dataset is about houses in Iowa! According to the Kaggle webpage, the competition is as follows:

> Ask a home buyer to describe their dream house, and they probably won’t begin with the height of the basement ceiling or the proximity to an east-west railroad. But this playground competition’s dataset proves that much more influences price negotiations than the number of bedrooms or a white-picket fence.
With 79 explanatory variables describing (almost) every aspect of residential homes in Ames, Iowa, this competition challenges you to predict the final price of each home.

More explicitly, our dataset has 81 columns, or features: 
1. `SalePrice`, our response variable $Y$ that we are trying to predict accurately and precisely 
2. `Id`, a simple identification variable
3. 79 explanatory variables $X_k$ that we can use to predict `SalePrice`. Some of the variables are categorical, and others are continuous quantitative. 

The goal of the next four weeks to to create a model that trains on (some of) the 79 explanators from the training set to predict `SalePrice` well in the test set. 

How will we know which explanators to use? We can start with some intuition and research into what each column represents by reading the data dictionary in `crash-course/Kaggle/DATA/house-prices/data_description.txt`. 

> Please take a moment to read over this dictionary, as you will need to have a keen sense of these features for the weeks ahead.

Now that we've talked a bit about this dataset, let's actually take a look at it. The first step is to load in the data! We will store this in a `pandas` dataframe.

In [1]:
import pandas as pd
train = pd.read_csv('DATA/house-prices/train.csv')
test = pd.read_csv('DATA/house-prices/test.csv')

#Let's see what the training dataframe looks like!
train.head(10)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   
5   6          50       RL         85.0    14115   Pave   NaN      IR1   
6   7          20       RL         75.0    10084   Pave   NaN      Reg   
7   8          60       RL          NaN    10382   Pave   NaN      IR1   
8   9          50       RM         51.0     6120   Pave   NaN      Reg   
9  10         190       RL         50.0     7420   Pave   NaN      Reg   

  LandContour Utilities    ...     PoolArea PoolQC  Fence MiscFeature MiscVal  \
0         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
1         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
2         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
3         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
4         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
5         Lvl    AllPub    ...            0    NaN  MnPrv        Shed     700   
6         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
7         Lvl    AllPub    ...            0    NaN    NaN        Shed     350   
8         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   
9         Lvl    AllPub    ...            0    NaN    NaN         NaN       0   

  MoSold YrSold  SaleType  SaleCondition  SalePrice  
0      2   2008        WD         Normal     208500  
1      5   2007        WD         Normal     181500  
2      9   2008        WD         Normal     223500  
3      2   2006        WD        Abnorml     140000  
4     12   2008        WD         Normal     250000  
5     10   2009        WD         Normal     143000  
6      8   2007        WD         Normal     307000  
7     11   2009        WD         Normal     200000  
8      4   2008        WD        Abnorml     129900  
9      1   2008        WD         Normal     118000  

[10 rows x 81 columns]

## Questions for Understanding:
> 1. How many features (columns) do we have? How many entries (rows)?
> 2. What does a single row represent in our dataset?
> 3. List 3 issues/questions that you see from the dataset. Some ideas to get started: What does `LotShape` represent? Is it a good thing that we have so many features? How does a large number of features affect our modeling approach?
> 4. What are some important things we should do for data cleaning and exploration? Some ideas to get started: are all the values in `LotFrontage` numeric? What about `Alley`? How should we fix missing/`NA` values that appear sporadically in some columns? What about columns that are almost entirely full of `NA` values? Some columns are qualitative strings, whereas others are qualitative numerics - how might this affect our cleaning?

# II. Data Cleaning
First, let's see how big our dataset looks like.

In [2]:
print("Training size",train.shape)
print("Test size",test.shape)

Training size (1460, 81)
Test size (1459, 80)


This tells us that we have $1460$ datapoints in the training set, and we would like to predict on $1459$ samples. There are $80$ total features, and one response variable we would like to predict. However, not all these 'features' are actually useful, such as `Id`. Thus we need to understand what the actual variables mean. A huge part in data science is actually understanding the variables. Of course it is possible to throw the data into some machine learning model and have it spit out predictions, but without actually understanding what data you are dealing with and how to feed the data into the model, your model is worthless.

Spend a good deal of time reading over the data dictionary. It is located in  
`crash-course/Kaggle/DATA/house-prices/data_description.txt`

## Questions for Understanding:
> 1. There are many categorical variables. What are some possibilities to deal with these variables? (Hint: [one hot encoding](https://www.kaggle.com/dansbecker/using-categorical-data-with-one-hot-encoding))
> 2. Are there any categorical variables that we can convert to numerical/quantitative variables as well? How might we do that?
> 3. Are there any variables that are just irrelevant and we can ignore?

## Dealing with NA values
In almost all datasets, we will have NA values. These can be a pain to deal with, as there are many viable choices of what to do. First, it is good to see what columns have NA values.

In [17]:
#Sum the number of NA's in each column
train.isnull().sum()

Id                  0
MSSubClass          0
MSZoning            0
LotFrontage       259
LotArea             0
Street              0
Alley            1369
LotShape            0
LandContour         0
Utilities           0
LotConfig           0
LandSlope           0
Neighborhood        0
Condition1          0
Condition2          0
BldgType            0
HouseStyle          0
OverallQual         0
OverallCond         0
YearBuilt           0
YearRemodAdd        0
RoofStyle           0
RoofMatl            0
Exterior1st         0
Exterior2nd         0
MasVnrType          8
MasVnrArea          8
ExterQual           0
ExterCond           0
Foundation          0
                 ... 
BedroomAbvGr        0
KitchenAbvGr        0
KitchenQual         0
TotRmsAbvGrd        0
Functional          0
Fireplaces          0
FireplaceQu       690
GarageType         81
GarageYrBlt        81
GarageFinish       81
GarageCars          0
GarageArea          0
GarageQual         81
GarageCond         81
PavedDrive

It seems that many of the features have a great deal of NAs. However, this is not necessarily the case. The astute reader will notice that some variables like `Alley`, `PoolQC`, and `Fence`, have NA has an actual value. For example for `Pool Quality`, we have these possibilities.

PoolQC: Pool quality
		
       Ex	Excellent
       Gd	Good
       TA	Average/Typical
       Fa	Fair
       NA	No Pool
So having NA values is not the usual not available, it can actually be a legitimate value! We need to parse through the data dictionary to see when NA's are actually significant, and when they actually mean NA.

## Questions for Understanding:
> 1. Go through the data dictionary and find all features that have NA as legitimate values.
> 2. How can we refactor the variables in question 1 appropriately?


The features that we found with NA's are:
`Alley`, `BsmtQual`, `BsmtCond`, `BsmtExposure`, `BsmtFinType1`, `BsmtFinType2`, `FireplaceQu`, `GarageType`, `GarageFinish`, `GarageQual`, `GarageCond`, `PoolQC`, `Fence`, `Misc Feature`.

Alley: NA for no alley access  
`BsmtQual`, `BsmtCond`, `BsmtExposure`, `BsmtFinType1`, `BsmtFinType2`: NA for no basement  
`FireplaceQu`: NA for no fireplace  
`GarageType`, `GarageFinish`, `GarageQual`, `GarageCond`: NA for no garage  
`PoolQC`: NA for no pool  
`Fence`: NA for no fence  
`Misc Feature`: NA for no other miscellanous features (i.e. elevator, 2nd garage, shed, tennis Court, other) 

## Data Cleaning Functions
We provide some functions to help with data cleaning and preprocessing. One is creating one hot encodings of various features, and the latter is converting a categorical feature into values.

### One Hot Encoding for Unordered Categorical Variables
Consider we have a feature `committee` for a data frame of SUSA members. Let's say for sake of simplicity there are four committees, `CX`, `DC`, `RP`, and `WD`. There is no inherent ordering to the features, but each member is only in one committee. Thus we can replace this categorical variable with a single boolean for `committee`. If a member is a part of `CX`, then they will have a $1$ for `DC` and $0$ for all other variables. We have written this function for you in `oneHotFeature`.

### Ordered Categorical Variables
Consider we have a feature `attendance` for a data frame of SUSA members. Let's say there are $3$ possible values, `Good`, `Ok`, `Poor`. These inherently have an ordering, `Good` is better than `Okay` which is better than `Poor`. We can assign a numerical value to these, such as $0$ for `Poor`, $1$ for `Ok`, and $2$ for `Good`. We have written this function for you in `categoricalToQuantitative`.


Please read over the following functions and read the comments carefully! They describe in detail what the functions do, and are **very** crucial to the data cleaning process.

In [18]:
# Data Cleaning Functions


# This function is for unordered categorical features 

# This function takes in input as: 
# Data frame 'df'
# List of features to one hot encode 'features'
# Boolean for how to deal with NA's 'withNA'

# This function creates a set of output features from a categorical feature
# The output features are one hot encodings with names featureName_{value}
# Returns a new data frame (does not modify original dataframe) with appended features
# Usually the NA values are also considered
# Will add a column featureName_none one hot encoding of NA values
# If the boolean withNA is false, it will not consider NA values

# If a feature is not found, it will ignore that feature and attempt to one hot the other features

# Code from Numpy and Pandas SUSA guide
# crash-course/Python/Numpy and Pandas.ipynb
def oneHotFeature(df, features, withNA=True):
    #Copy over data
    newDf = df.copy()
    for feature in features:
        try:
            if withNA:
                newDf[feature] = newDf[feature].fillna('none')
            col_onehot = pd.get_dummies(newDf[feature], prefix=feature) 
            newDf.drop(feature, axis=1)
            newDf = newDf.join(col_onehot)
        except:
            print("No such feature", feature, "found in the dataframe when trying to one-hot encode!")
    return newDf


# This function is for ordered categorical features 

# This function takes in input as: 
# Data frame 'df'
# A single categorical feature to to be mapped 'feature'
# A mapping dictionary 'mapping'

# This function creates takes a dataframe and categorical feature, and maps the categorical values
# using the dictionary mapping
# Returns a new data frame (does not modify original dataframe) with modified values for the feature column
# For mappings with NA values, use 'NA' in the dictionary, this function properly deals with them

# If the mapping is just from 0 to n-1 for n values
# Then set default to True, and mapping is instead a list of ordering from worst to best

# If a feature is not found, the function will fail
def categoricalToQuantitative(df,feature, mapping,assumeInOrder = False):
    newDf = df.copy()
    try:
        newDf[feature] = newDf[feature].fillna('none')
        if assumeInOrder:
            newMapping ={}
            for i in range(len(mapping)):
                if mapping[i] == 'NA':
                    newMapping['none'] = i 
                else:
                    newMapping[mapping[i]] = i
        else:    
            newMapping = mapping.copy()
            if 'NA' in newMapping.keys():
                newMapping['none'] = newMapping['NA']
        newDf[feature] = newDf[feature].apply(lambda feat: newMapping[feat] if feat in newMapping.keys() else feat)
    except:
        print("No such feature", feature, "found in the dataframe when trying to map!")
    return newDf

Let's use the first function `oneHotFeature` first. Consider the `Heating` feature. There are $5$ possible values, Brick & Tile, Cinder Block, Poured Concrete, Slab, Stone and Wood. There isn't an inherent ordering to these, where one is better than the other. The best way to go about preprocessing the data is through one hot encoding the data. We remove the `Foundation` feature and replace it with $5$ separate boolean features, one for each of the possible values.

The dataframe is shown below. Keep in mind that the original dataframe, `train`, is not modified in this function call.

In [19]:
oneHotFeature(train,['Foundation']).head(10)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   
5   6          50       RL         85.0    14115   Pave   NaN      IR1   
6   7          20       RL         75.0    10084   Pave   NaN      Reg   
7   8          60       RL          NaN    10382   Pave   NaN      IR1   
8   9          50       RM         51.0     6120   Pave   NaN      Reg   
9  10         190       RL         50.0     7420   Pave   NaN      Reg   

  LandContour Utilities       ...        YrSold SaleType SaleCondition  \
0         Lvl    AllPub       ...          2008       WD        Normal   
1         Lvl    AllPub       ...          2007       WD        Normal   
2         Lvl    AllPub       ...          2008       WD        Normal   
3         Lvl    AllPub       ...          2006       WD       Abnorml   
4         Lvl    AllPub       ...          2008       WD        Normal   
5         Lvl    AllPub       ...          2009       WD        Normal   
6         Lvl    AllPub       ...          2007       WD        Normal   
7         Lvl    AllPub       ...          2009       WD        Normal   
8         Lvl    AllPub       ...          2008       WD       Abnorml   
9         Lvl    AllPub       ...          2008       WD        Normal   

  SalePrice Foundation_BrkTil Foundation_CBlock Foundation_PConc  \
0    208500                 0                 0                1   
1    181500                 0                 1                0   
2    223500                 0                 0                1   
3    140000                 1                 0                0   
4    250000                 0                 0                1   
5    143000                 0                 0                0   
6    307000                 0                 0                1   
7    200000                 0                 1                0   
8    129900                 1                 0                0   
9    118000                 1                 0                0   

   Foundation_Slab  Foundation_Stone  Foundation_Wood  
0                0                 0                0  
1                0                 0                0  
2                0                 0                0  
3                0                 0                0  
4                0                 0                0  
5                0                 0                1  
6                0                 0                0  
7                0                 0                0  
8                0                 0                0  
9                0                 0                0  

[10 rows x 87 columns]

In [20]:
# Train is not modified
train['Foundation'].head(10)

0     PConc
1    CBlock
2     PConc
3    BrkTil
4     PConc
5      Wood
6     PConc
7    CBlock
8    BrkTil
9    BrkTil
Name: Foundation, dtype: object

Now we may try this with the other function `categoricalToQuantitative`, mapping a categorical data where there is an quantitative correspondence. Consider the feature `FireplaceQu`. There are $6$ possible values, `Ex`, `Gd`, `TA`, `Fa`, `Po`, `NA`, corresponding to Excellent, Good, Average, Fair, Poor, and No fireplace. These could be mapped to the values from $0$ to $5$, in order of quality. Thus we can make a dictionary of values, where `Ex` corresponds to $5$, `Gd` corresponds to $4$, ..., `NA` corresponds to zero. This is written out below.

In [21]:
mapping = {}
mapping['NA'] = 0
mapping['Po'] = 1
mapping['Fa'] = 2
mapping['TA'] = 3
mapping['Gd'] = 4
mapping['Ex'] = 5

print(train['FireplaceQu'].head(10))
print(categoricalToQuantitative(train,'FireplaceQu',mapping)['FireplaceQu'].head(10))

0    NaN
1     TA
2     TA
3     Gd
4     TA
5    NaN
6     Gd
7     TA
8     TA
9     TA
Name: FireplaceQu, dtype: object
0    0
1    3
2    3
3    4
4    3
5    0
6    4
7    3
8    3
9    3
Name: FireplaceQu, dtype: int64


However, it is a bit annoying to type out the entire dictionary like that if we would like to order from $0$ to $5$. Another variable in `categoricalToQuantitative` is `default`. If `assumeInOrder` is true, then mapping is instead a list of variables rather than a dictionary, and we can just pass in the order of values from worst to best. This maps from $0$ to $n-1$ when there are $n$ possible values. This is more convenient, but it may be better at times to be able to customize how you want to map the values.

In [22]:
mapping = ['NA','Po','Fa','TA','Gd','Ex']
#Notice these mappings are the same
print(categoricalToQuantitative(train,'FireplaceQu',mapping,assumeInOrder=True)['FireplaceQu'].head(10))

0    0
1    3
2    3
3    4
4    3
5    0
6    4
7    3
8    3
9    3
Name: FireplaceQu, dtype: int64


If you would like to remove a feature, use the code below.

In [23]:
dataframe.drop(feature, axis=1)

NameError: name 'dataframe' is not defined

Use the above two functions to clean the dataset! This may take a while, but doing a good job should take a while. Decide what variables are not worth keeping, decide what categorical features need to be changed, and how they should be changed. Consider how to deal with NA values, and keep all these commands together. We would recommend to save the final cleaned file as a csv so that you may easily reopen and send it, and also keep all the commands together neatly in the code block below.

Some recommendations:
> 1. Convert all features into quantitative values
> 2. While cleaning, keep in mind some features that you feel are very helpful.
> 3. Remove features that do not seem important.

In [25]:
# DATA CLEANING
# To start off data cleaning
clean = train.copy()


clean = categoricalToQuantitative(clean,'FireplaceQu',['NA','Po','Fa','TA','Gd','Ex'],assumeInOrder=True)
clean = oneHotFeature(clean,['Foundation'])

# DATA CLEANING SOLUTION

# ====================== DO NOT INCLUDE IN FINAL WORKBOOK ========================

unorderedFeatures = ['MSZoning','Street','Alley','LandContour','LotConfig','Neighborhood']
clean = oneHotFeature(clean,unorderedFeatures)
clean = categoricalToQuantitative(clean,'LotShape',['IR3','IR2','IR1','R'],assumeInOrder=True)
clean = categoricalToQuantitative(clean,'Utilities',['ELO','NoSeWa','NoSewr','AllPub'],assumeInOrder=True)
clean = categoricalToQuantitative(clean,'LandSlope',['Sev','Mod','Gtl'],assumeInOrder=True)
clean = categoricalToQuantitative(clean,'LandSlope',['Sev','Mod','Gtl'],assumeInOrder=True)

## STILL IN PROGRESS

# ================================================================================

clean.head(10)

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave  none      Reg   
1   2          20       RL         80.0     9600   Pave  none      Reg   
2   3          60       RL         68.0    11250   Pave  none        2   
3   4          70       RL         60.0     9550   Pave  none        2   
4   5          60       RL         84.0    14260   Pave  none        2   
5   6          50       RL         85.0    14115   Pave  none        2   
6   7          20       RL         75.0    10084   Pave  none      Reg   
7   8          60       RL          NaN    10382   Pave  none        2   
8   9          50       RM         51.0     6120   Pave  none      Reg   
9  10         190       RL         50.0     7420   Pave  none      Reg   

  LandContour  Utilities          ...          Neighborhood_NoRidge  \
0         Lvl          3          ...                             0   
1         Lvl          3          ...                             0   
2         Lvl          3          ...                             0   
3         Lvl          3          ...                             0   
4         Lvl          3          ...                             1   
5         Lvl          3          ...                             0   
6         Lvl          3          ...                             0   
7         Lvl          3          ...                             0   
8         Lvl          3          ...                             0   
9         Lvl          3          ...                             0   

   Neighborhood_NridgHt Neighborhood_OldTown Neighborhood_SWISU  \
0                     0                    0                  0   
1                     0                    0                  0   
2                     0                    0                  0   
3                     0                    0                  0   
4                     0                    0                  0   
5                     0                    0                  0   
6                     0                    0                  0   
7                     0                    0                  0   
8                     0                    1                  0   
9                     0                    0                  0   

  Neighborhood_Sawyer Neighborhood_SawyerW Neighborhood_Somerst  \
0                   0                    0                    0   
1                   0                    0                    0   
2                   0                    0                    0   
3                   0                    0                    0   
4                   0                    0                    0   
5                   0                    0                    0   
6                   0                    0                    1   
7                   0                    0                    0   
8                   0                    0                    0   
9                   0                    0                    0   

   Neighborhood_StoneBr  Neighborhood_Timber  Neighborhood_Veenker  
0                     0                    0                     0  
1                     0                    0                     1  
2                     0                    0                     0  
3                     0                    0                     0  
4                     0                    0                     0  
5                     0                    0                     0  
6                     0                    0                     0  
7                     0                    0                     0  
8                     0                    0                     0  
9                     0                    0                     0  

[10 rows x 131 columns]

To save the dataframe as csv, run this line of code.

In [143]:
# Save to csv
clean.to_csv('DATA/house-prices/train_cleaned.csv')

# III. Exploratory Data Analysis

Now that we've 

## Data Visualization

# IV. Modeling

## Feature Selection

## A First Approach to Machine Learning: Linear Regression

# V. Model Evaluation

Now it's time to actually see how your model performed

# Extensions to Linear Regression

# Conclusion

This ends our textbook-style primer into deep learning with Keras. While this was just an introduction to neural nets, we hope that you can now see some of the workflow patterns associated with machine learning. Feel free to play around with the code above to get a better feel for the hyperparameters of the neural net model. As always, please email [`contact@arun.run`](mailto:contact@arun.run) or [`prc@berkeley.edu`](mailto:prc@berkeley.edu) with any questions or concerns whatsoever. Happy machine learning!

## Sneakpeek at SUSA Kaggle Competition II

After Spring Break, we will be guiding you through a four-week collaborative Kaggle competition with your peers in Career Exploration! We want to give you the experience of working with real data, using real machine learning algorithms, in an educational setting. You will have to choose either Python or R, and dive into reading kernels on the Kaggle website, use visualization and feature engineering to improve your score, and maybe even pick up a few advanced deep learning models along the way. If this sounds a bit intimidating right now, do not fret! Your SUSA Mentors will be there to mentor you through the whole thing. So rest up during Spring Break, and come back ready to tackle your biggest data challenge yet!

# Additional Reading
* For more information on the Kaggle API, a command-line program used to download and manage Kaggle datasets, visit the [Kaggle API Github page](https://github.com/Kaggle/kaggle-api)  
* For an interactive guide to learning R and Python, visit [DataCamp](https://www.datacamp.com/) a paid tutorial website for learning data computing.
